content/blog/2019-12-06-alive/index.md

```diff
@@ -0,0 +1,181 @@
+
+## Grokking undefined behavior
+The greatest technical challenge for a compiler or verification engineer in this space is wrangling with undefined behavior.
+One of the authors of Alive, John Regehr, has [several][jr1] [excellent][jr2] [blog][jr2] posts on the topic.
+
+### Refinement
+By definition, compilers are allowed to produce different results for the same source program in the presence of undefined behavior.
+However, compilers are _not_ allowed to introduce undefined behavior for a program and input that was well-defined in the unoptimized source code.
+That is, the burden on a verifier like Alive is to show that optimization targets are _refinements_ of the source: the optimized target can include a more, but not less, specific subset of behaviors of the source.
+
+To illustrate this, let's look at a [real bug][PR20186] in an optimization that Alive discovered in production LLVM (also described [here][jr1]).
+The optimization aims to simplify an expression that negates the division of a variable `x` with a constant `C`, from the explicit `0 - (x/C)`, to the simpler `x / -C`.
+
+In the Alive DSL, we specify this with:
+
+```
+%div = sdiv %x, C
+%r = sub 0, %div
+  =>
+%r = sdiv %x, -C
+```
+
+When we hand this optimization off to Alive, we get:
+
+```
+Precondition: true
+%div = sdiv %x, C
+%r = sub 0, %div
+  =>
+%r = sdiv %x, -C
+
+Done: 1
+ERROR: Domain of definedness of Target is smaller than Source's for i2 %r
+
+Example:
+%x i2 = 2 (0x2)
+C i2 = 1 (0x1)
+%div i2 = 2 (0x2)
+Source value: 2 (0x2)
+Target value: undef
+```
+
+The problem here is the interplay between an edge case (signed integer overflow) and undefined behavior.
+When the concrete type is `i2` and the values are `x = -2` and `C = 1`, `x/-C = -2/-1 = 2`, but `2` overflows a 2-bit signed integer! While mathematically this is also true in the source template, LLVM's language reference states that overflow in `sdiv` is undefined behavior, the same of which is not true for `sub`.
+Thus, the target template introduced undefined behavior in a case where there previously was none, so it is _not_ a refinement.
+
+In order to fix this bug, the LLVM developers added a precondition that `C != 1` and `C` is not a sign bit.
+In Alive, we can represent this precondition as ` ((C != 1) && !isSignBit(C))`, and the optimization verifies.
+
+### Poison
+An additional complication in handling undefined behavior is that LLVM actually has two flavors of _deferred_ (non-crashing) undefined behavior: the `undef` value, and implicit _poisoned_ values.
+
+Poison values are a stronger form of undefined behavior: they happen when a side-effect-free instruction produces a result that might later trigger undefined behavior.
+The true undefined behavior only occurs if/when a poisoned value is later used by an instruction that _does_ have side effects (for example, a division by zero).
+Poison values are not represented explicitly in LLVM IR, and can only be identified via careful analysis.
```

Comment:
In case you're curious, they even wrote a whole paper about distinguishing between undef and poison:
https://www.cs.utah.edu/~regehr/papers/undef-pldi17.pdf

And a concrete proposal to replace them both with more precise poison-like semantics:
https://lists.llvm.org/pipermail/llvm-dev/2016-October/106182.html

…which IIRC is finally making some implementation progress recently?

[Edit: Oops, noticed you linked to the latter already below!]

Reply:
I also link to the blog post of the first paper in the first series of John Regehr links!